### PR TITLE
Fix to undefined fraction denominator: Improved testing and better edge case handling

### PR DESCRIPTION
**Pull request by undermark5**
This PR is a combination of #3 plus the changes that I reference in [my comment](https://github.com/nsafai/recipe-parser/pull/3#issuecomment-787506900).

The biggest changes are an increase in the number of test cases (I'm not sure if they are in the proper place code wise or implemented properly) and the ability to handle more complex inputs including fractions (e.g. `"1-2 1/2 14.5 oz can tomato paste"` -> `{quantity: '1-2.5', unit: '14.5 ounce', ingredient: 'can tomato paste', minQty: '1', maxQty: '2.5'}`)

The multiplication is definitely an option, and maybe that could be a configuration option (or optional parameter), but I feel like people have an easier time doing multiplication than division.

### CODE REVIEW

File: test/index.test.ts [hunk 1]
```diff
@@ -219,6 +219,66 @@ describe('recipe parser', () => {
     });
   });
 
+  it('doesn\'t explode when 1 2 lb. chicken', () => {
+    expect(parse('1 2 lb. chicken')).to.deep.equal({
+      unit: '2 pound',
```
Comment (nsafai) [inline test/index.test.ts]: Hmm, I'm a bit concerned about this because 2 isn't really a "unit of measurement"

For example, what happens if someone uses this recipe parser to combine ingredients from multiple recipes to create a shopping list? If >2 recipes have chicken in it, how will it be able to combine: 
``` ts
{
      unit: '2 pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
with
```ts
{
      unit: 'pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
 ```
 
Another edge case this doesn't cover is for use-cases where the units need to be converted.
How do you convert `"2 pound" => kg`?

Comment (nsafai) [reply test/index.test.ts]: I wonder if it be safer to omit a unit in this case? e.g.
```ts
{
      unit: null,
      ingredient: '2 lb. chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```

**Review by nsafai — COMMENTED** [review]
Hey @undermark5 thanks for raising the PR. I do have some concerns with making this change because it can limit usability of the `combine()` feature and the future possibility of being able to convert units. Would love to get your thoughts

File: test/index.test.ts (hunk 1 above)
Comment (undermark5) [reply test/index.test.ts]: Ya, I understand the concern because a numeric value isn't really a unit of measure. However, I think the suggestion to not have a unit is actually worse than having a numeric part of the unit.

I haven't taken a look at the logic for combining ingredients, but I think it should be simple enough to do. First step would be assume that any unit without a numeric part has a 1 for the numeric part, the second would then be to convert units to the same base unit (adjusting any associated numeric portion correctly), finally sum up all numeric portions (including the assumed 1's) and then recombine. Which means that combining
```ts
{
      unit: '2 pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
with
```ts
{
      unit: 'pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
will result in
```ts
{
      unit: '3 pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
However, an issue comes into play when we have a quantity (and quantity ranges make it even worse). Should the number in the unit be multiplied by the quantity? Sometimes that is fine, but other times it doesn't make sense (in the case of cans or other uniform discrete items).

Maybe it would be better to take the number out of the unit and change the quantity accordingly. However, this still isn't perfect either. For example combining
```ts
{
      unit: '10.5 oz can',
      ingredient: 'cream of chicken soup',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
with 
```ts
{
      unit: '10.5 oz can',
      ingredient: 'cream of chicken soup',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
should not result in 
```ts
{
      unit: 'oz can',
      ingredient: 'cream of chicken soup',
      quantity: '20.1',
      minQty: '20.1',
      maxQty: '20.1'
}
```
Though I think we might be safe to say that if there is a unit of measure (eg oz, tbsp, lb) as well as an additional noun, we should only combine them if the unit matches exactly and simply add the quantities together. Following the logic above with the example of chicken we have 
```ts
{
      unit: '2 pound', //this actually would actually be 'pound' and the quantities would all be 2 times what they currently are
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
with
```ts
{
      unit: 'pound',
      ingredient: 'chicken',
      quantity: '1',
      minQty: '1',
      maxQty: '1'
}
```
will result in
```ts
{
      unit: 'pound',
      ingredient: 'chicken',
      quantity: '3',
      minQty: '3',
      maxQty: '3'
}
```
Hopefully that makes some sense?

I actually should have some time to work on this further this week if you would like to see an implementation of my proposal.

Comment (undermark5) [reply test/index.test.ts]: I think my exception for when there is a noun and a unit of measure in the unit field works out fine for combining logic, and conversion from pound -> kg or oz -> ml (or oz -> g) won't be too difficult as we use the number in the unit instead of the quantity.

Comment (undermark5) [reply test/index.test.ts]: Ah, looking at it now my idea won't work, because there isn't ever multiple words in the unit, and it is incredibly difficult if not impossible to determine if a word after a unit of measure should be part of that "unit" or part of the ingredient.

Maybe everything I said above is beyond what is possible and not having any unit might actually be the better option now that I look at it again.
